wear/src/main/java/io/homeassistant/companion/android/complications/EntityStateDataSourceService.kt

```diff
@@ -0,0 +1,86 @@
+package io.homeassistant.companion.android.complications
+
+import android.graphics.Color
+import android.graphics.drawable.Icon
+import androidx.wear.watchface.complications.data.ComplicationData
+import androidx.wear.watchface.complications.data.ComplicationType
+import androidx.wear.watchface.complications.data.MonochromaticImage
+import androidx.wear.watchface.complications.data.PlainComplicationText
+import androidx.wear.watchface.complications.data.ShortTextComplicationData
+import androidx.wear.watchface.complications.datasource.ComplicationRequest
+import androidx.wear.watchface.complications.datasource.SuspendingComplicationDataSourceService
+import com.mikepenz.iconics.IconicsColor
+import com.mikepenz.iconics.IconicsDrawable
+import com.mikepenz.iconics.typeface.library.community.material.CommunityMaterial
+import com.mikepenz.iconics.utils.backgroundColor
+import com.mikepenz.iconics.utils.colorInt
+import dagger.hilt.android.AndroidEntryPoint
+import io.homeassistant.companion.android.common.R
+import io.homeassistant.companion.android.common.data.integration.IntegrationRepository
+import io.homeassistant.companion.android.common.data.integration.domain
+import io.homeassistant.companion.android.database.wear.EntityStateComplicationsDao
+import io.homeassistant.companion.android.util.getIcon
+import javax.inject.Inject
+
+@AndroidEntryPoint
+class EntityStateDataSourceService : SuspendingComplicationDataSourceService() {
+
+    @Inject
+    lateinit var integrationUseCase: IntegrationRepository
+    @Inject
+    lateinit var entityStateComplicationsDao: EntityStateComplicationsDao
+
+    override suspend fun onComplicationRequest(request: ComplicationRequest): ComplicationData? {
+        if (request.complicationType != ComplicationType.SHORT_TEXT)
+            return null
+
+        val id = request.complicationInstanceId
+
+        val entityId = entityStateComplicationsDao.get(id)?.entityId
+            ?: return ShortTextComplicationData.Builder(
+                text = PlainComplicationText.Builder(getText(R.string.complication_entity_invalid)).build(),
+                contentDescription = PlainComplicationText.Builder(getText(R.string.complication_entity_state_content_description))
+                    .build()
+            ).build()
+
+        val entity = integrationUseCase.getEntity(entityId)
```

Comment:
An integration request also might throw an exception, maybe catch it + return `null` to indicate no update instead of crashing? Or doesn't the system mind uncaught exceptions and will the complication keep working?

Reply:
Don't know if it cares about uncaught exceptions, but indeed simple change to make it catch it. Good call ;)